service/controller/clusterapi/v27/resource/ipam/create.go

```diff
@@ -144,24 +123,36 @@ func (r *Resource) allocateSubnet(ctx context.Context) (net.IPNet, error) {
 
 	err = g.Wait()
 	if err != nil {
-		return net.IPNet{}, microerror.Mask(err)
+		return nil, microerror.Mask(err)
 	}
 
 	reservedSubnets = ipam.CanonicalizeSubnets(r.networkRange, reservedSubnets)
 
-	var subnet net.IPNet
-	{
-		r.logger.LogCtx(ctx, "level", "debug", "message", "finding free subnet")
+	return reservedSubnets, nil
+}
 
-		subnet, err = ipam.Free(r.networkRange, r.allocatedSubnetMask, reservedSubnets)
-		if err != nil {
-			return net.IPNet{}, microerror.Maskf(err, "networkRange: %s, allocatedSubnetMask: %s, reservedSubnets: %#v", r.networkRange.String(), r.allocatedSubnetMask.String(), reservedSubnets)
-		}
+func (r *Resource) splitAndPersistReservedSubnet(ctx context.Context, cr v1alpha1.AWSConfig, subnet net.IPNet, azs []string) error {
+	statusAZs, err := splitSubnetToStatusAZs(subnet, azs)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "updating CR status to persist network allocation and chosen availability zones")
 
-		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found free subnet %#q", subnet.String()))
+	cr.Status.Cluster.Network.CIDR = subnet.String()
+	cr.Status.AWS.AvailabilityZones = statusAZs
+
+	_, err = r.g8sClient.ProviderV1alpha1().AWSConfigs(cr.Namespace).UpdateStatus(&cr)
+	if err != nil {
+		return microerror.Mask(err)
 	}
 
-	return subnet, nil
+	r.logger.LogCtx(ctx, "level", "debug", "message", "updated CR status to persist network allocation and chosen availability zones")
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "canceling reconciliation")
+	reconciliationcanceledcontext.SetCanceled(ctx)
```

Comment:
IMO we should not use control flow primitives of the operator framework in some private methods. It is not obvious that the flow is controlled here from the outside and makes it only harder to comprehend how that thing works. It would be good to only use these primitives in resource interface methods like `EnsureCreated`, but not deeper in the code.

Reply:
Good shout. I'll fix these.